lib/line_offsets.ml

```diff
@@ -0,0 +1,84 @@
+
+(* A very basic dynamic array. *)
+module Vec = struct
+  type 'a t = {
+    mutable arr: 'a array;
+    mutable sz: int;
+  }
+
+  let create() : _ t = { arr=[||]; sz=0 }
+
+  let grow_ self x =
+    let new_size = max 6 (let n = Array.length self.arr in n + n / 2 + 1) in
+    let arr = Array.make new_size x in
+    Array.blit self.arr 0 arr 0 self.sz;
+    self.arr <- arr
+
+  let push (self:_ t) x =
+    if self.sz = Array.length self.arr then grow_ self x;
+    self.arr.(self.sz) <- x;
+    self.sz <- 1 + self.sz
+
+  let to_array self = Array.sub self.arr 0 self.sz
+end
+
+type t = int array
+
+let from_string (s:string) : t =
+  let lines = Vec.create() in
+  Vec.push lines 0; (* first line is free *)
+  let size = String.length s in
+  let i = ref 0 in
+  while !i < size do
+    match String.index_from_opt s !i '\n' with
+    | None -> i := size
+    | Some j ->
+      Vec.push lines j;
+      i := j+1;
+  done;
+  Vec.to_array lines
+
+let from_chan (cin:in_channel) : t =
+  let buf = Buffer.create 1024 in
+  (try while true do Buffer.add_channel buf cin (16 * 1024) done; assert false
+   with End_of_file -> ());
+  from_string (Buffer.contents buf)
```

Comment:
I think you need to explicitly seek to position 0 of the channel before starting to read.
Otherwise, this will fail when calling two successive `pp` on the same input (created from an `in_channel`): the first pp will consume the input, and the second pp will not be able to compute line offsets because of that.